sql/sql_class.cc

```diff
@@ -5604,6 +5604,8 @@ void THD::leave_locked_tables_mode()
       mysql_ha_set_explicit_lock_duration(this);
     if (ull_hash.records)
       mysql_ull_set_explicit_lock_duration(this);
+    if (mdl_backup_lock)
+      mdl_context.set_lock_duration(mdl_backup_lock, MDL_EXPLICIT);
```

Comment:
There's broader problem, please see a comment around https://github.com/MariaDB/server/blob/10.4/sql/lock.cc#L998

In other words FTWRL is expecting connection to hold no meta-data locks, otherwise it may fall into a deadlock which cannot be ruled out even by the mdl deadlock detector. An example of such deadlock is described in https://jira.mariadb.org/browse/MDEV-20946.

The only good fix I can think of is to allow `BACKUP LOCK` only under `BACKUP STAGE.

Reply:
@svoj thank you for the comment! Do you mean that the fix introduces the new deadlock scenarios or are you about generic solution for MDEV-20946, not this one? We didn't find any new deadlock scenarios introduced with the fix.

We love the proposal
>  to allow `BACKUP LOCK` only under `BACKUP STAGE`
to fix MDEV-20946 - the lock is intended only for backups (I think). But it seems this change will affect the user API. We reviewed https://mariadb.com/kb/en/backup-lock/ and https://mariadb.com/kb/en/backup-stage/ - it seems the change won't break anything which was described as allowed by the KBs, but the locking scenarios from MDEV-20945 and MDEV-20946 won't work. Is such change allowed for 10.5 and earlier releases?

@vuvova could you please comment on this? Are we good to go with the change not only for 10.6, but for current 10.5 and stable 10.4?

Reply:
> Do you mean that the fix introduces the new deadlock scenarios 
@krizhanovsky, no, I mean the change introduced by this PR will likely become a dead-code when the broader problem is fixed. There're no new deadlocks introduced by the fix for sure.

By the "behaviour change" you mean not being able to issue `BACKUP LOCK` out of `BACKUP STAGE`? IMHO it should be fine.

I'm unsure if @vuvova has followed all this backup effort, it'd be nice if @montywi could comment on this.

Reply:
Yes, @svoj is right, I didn't follow it that much, better ask @montywi 

But https://mariadb.com/kb/en/backup-lock/ does say, though
> One should use different connections for BACKUP STAGE commands and BACKUP LOCK. 

which implies that you cannot only allow `BACKUP LOCK` under `BACKUP STAGE`

Reply:
Gosh, I missed that. We need to figure out the reasoning, I can't see any immediately. @vlad-lesin is `mariabackup` making use of `BACKUP LOCK`. If yes, are those called in a separate connection? At which `BACKUP STAGE`?

Reply:
Assuming `BACKUP STAGE` family is used, is there a need to use `BACKUP LOCK`? There is already `BACKUP STAGE BLOCK_DDL` that should block DDL. `BACKUP LOCK` looks redundant in such scenario.

`BACKUP LOCK` itself looks very strange. https://mariadb.com/kb/en/backup-lock/ assumes that data copying should be performed after `BACKUP UNLOCK`. Wouldn't copying the data without any locks cause a potential data corruption? There could be writes to the data store, since nothing prevents that.

Reply:
@i-rinat, IIRC InnoDB tables are being copied during `BACKUP STAGE START` when DDL is permitted. `BACKUP LOCK` was supposed to keep .frm and InnoDB data file in sync.

I don't completely understand the need for this statement either given that we have DDL logger.

Reply:
> IIRC InnoDB tables are being copied during BACKUP STAGE START when DDL is permitted. BACKUP LOCK was supposed to keep .frm and InnoDB data file in sync.

So it's just another way to backup InnoDB tables? Besides `BACKUP STAGE`? Latter seems like a completed feature set, while `BACKUP LOCK` is a tool for a narrow specific task of backuping InnoDB tables?

Sorry, I just can't wrap my head around it. If somebody uses `BACKUP STAGE`, why would they use BACKUP LOCK at all? [kb/en/mariabackup-and-backup-stage-commands](https://mariadb.com/kb/en/mariabackup-and-backup-stage-commands/#mariabackup-and-backup-stage-commands-in-mariadb-enterprise-server) describes backup steps. They look self-sufficient, with no additional locking needed. So `BACKUP LOCK` would be of no use in `BACKUP STAGE` family. If it is to be restricted to be called only inside `BACKUP STAGE`'s, then `BACKUP LOCK` may as well be removed completely.

Reply:
The usage of backup lock is, as said above, to keep .frm and table files in sync.
The ddl log is not usable to detect this case as we use to .frm to check with the ddl log if
the table was copied before or after the ddl.  If all table files are not in sync, then mariabackup could think it got the latest version of the tables, which it didn't

Reply:
It's correct that we do unlock before copying the tables. This works as we open the table BEFORE doing an unlock. If there is any alter table while we are copying the tables, it will not affect the backup as we are working with the original tables

Reply:
I have updated MDEV-20945 how to do a proper fix.
Can you please do a new patch based on the given suggestions?

Reply:
> do a new patch based on the given suggestions?

I'll make the changes.

Reply:
> do a new patch based on the given suggestions

Updated the patch.
It was a bit easier to also forbid data modifications too. Should be fine though, as it's unlikely that anybody would want to insert data while under BACKUP LOCK.